src/sizes/retained_size_by_classloaders.cpp

```diff
@@ -0,0 +1,102 @@
+    jvmtiError err = JVMTI_ERROR_NONE;
+    std::vector<jclass> filteredClasses;
+    for (jsize i = 0; i < *cnt; i++) {
+        jobject rootClassLoader = getClassLoader(env, classes[i]);
+        if (!env->IsSameObject(rootClassLoader, NULL)) {
+            for (jsize j = 0; j < env->GetArrayLength(classLoadersArray); j++) {
+                if (isEqual(env, env->GetObjectArrayElement(classLoadersArray, j), rootClassLoader)) {
+                    filteredClasses.emplace_back(classes[i]);
+                    break;
+                }
+            }
+        }
+    }
+
+    std::vector<jclass> nonfilteredClasses;
+    for (jsize i = 0; i < *cnt; i++) {
+        if (std::find(filteredClasses.begin(), filteredClasses.end(), classes[i]) == filteredClasses.end()) {
+            nonfilteredClasses.emplace_back(classes[i]);
+        }
+    }
+    std::cout << "SIZE: " << nonfilteredClasses.size() << std::endl;
+
+    err = tagObjectsOfClassLoaderClasses(toJavaArray(env, nonfilteredClasses), env->GetArrayLength(classLoadersArray));
+}
+
+jlong RetainedSizeByClassLoadersAction::getHeapSize(){
+    jlong totalSize = 0;
+    IterateThroughHeap(0, nullptr, countHeapSize, &totalSize,
+                       "calculate heap size");
+    return totalSize;
+}
+
+RetainedSizeByClassLoadersAction::RetainedSizeByClassLoadersAction(JNIEnv *env, jvmtiEnv *jvmti, jobject object)
+        : RetainedSizeAction(env, jvmti, object) {
+
+}
+
+jvmtiError RetainedSizeByClassLoadersAction::getRetainedSizeByClassLoaders(jobjectArray classLoadersArray,
+                                                                           std::vector <jlong> &result) {
+    jvmtiError err = JVMTI_ERROR_NONE;
+    jclass *classes;
+    jint cnt;
+    err = jvmti->GetLoadedClasses(&cnt, &classes);
+    std::cout << cnt << std::endl;
+    for (jsize i = 0; i < env->GetArrayLength(classLoadersArray); i++) {
+        jvmtiError err = tagObjectsByClassLoader(classes, &cnt, env->GetObjectArrayElement(classLoadersArray, i), i);
+    }
+    tagOtherObjects(classes, &cnt, classLoadersArray);
+
+    if (!isOk(err)) return err;
+    if (shouldStopExecution()) return MEMORY_AGENT_INTERRUPTED_ERROR;
+
+    err = tagHeap();
+    if (!isOk(err)) return err;
+    if (shouldStopExecution()) return MEMORY_AGENT_INTERRUPTED_ERROR;
+
+    result.resize(env->GetArrayLength(classLoadersArray) + 1);
+    err = IterateThroughHeap(JVMTI_HEAP_FILTER_UNTAGGED, nullptr, visitObject, result.data(),
+                             "calculate retained size");
+    return err;
+}
+
+jlongArray RetainedSizeByClassLoadersAction::executeOperation(jobjectArray classLoadersArray) {
+    std::vector <jlong> result;
+    jvmtiError err = getRetainedSizeByClassLoaders(classLoadersArray, result);
+    if (!isOk(err)) {
+        handleError(jvmti, err, "Could not estimate retained size by classLoaders");
+        return env->NewLongArray(0);
+    }
+    std::cout << "HEAP SIZE: " << getHeapSize() << std::endl;
```

Comment:
Since memory agent is a library, it would be better not to pollute stdout, but instead return the result in the format of nested arrays: `[[class loaders sizes as jlongs], [other classes size]]`